source/common/quic/envoy_quic_client_stream.cc

```diff
@@ -57,7 +57,11 @@ Http::Status EnvoyQuicClientStream::encodeHeaders(const Http::RequestHeaderMap&
       sent_head_request_ = true;
     }
   }
+  uint64_t bytes_written = stream_bytes_written();
```

Comment:
stream_bytes_written() counts the bytes written to the wire. However, calling WriteHeaders() doesn't guarantee the serialized headers will go to the wire immediately.  In addition to stream_bytes_written(), I suspect we need to monitor buffered bytes in send buffer as well.